utils/resources.py

```diff
@@ -0,0 +1,101 @@
+import json
+import os
+import shutil
+import sublime
+import subprocess
+import threading
+
+from sublime_lib import ResourcePath
+
+
+def run_command(on_exit, on_error, popen_args):
+    """
+    Runs the given args in a subprocess.Popen, and then calls the function
+    on_exit when the subprocess completes.
+    on_exit is a callable object, and popen_args is a list/tuple of args that
+    on_error when the subprocess throws an error
+    would give to subprocess.Popen.
+    """
+    def run_in_thread(on_exit, on_error, popen_args):
+        try:
+            subprocess.check_call(popen_args, shell=sublime.platform() == 'windows')
+            on_exit()
+        except subprocess.CalledProcessError as error:
+            on_error(error)
+
+    thread = threading.Thread(target=run_in_thread, args=(on_exit, on_error, popen_args))
+    thread.start()
+    # returns immediately after the thread starts
+    return thread
+
+
+def log_and_show_message(msg, additional_logs=None):
+    print(msg, '\n', additional_logs) if additional_logs else print(msg)
+    sublime.active_window().status_message(msg)
+
+
+class ServerNpmResources(object):
+    """Global object providing paths to server resources.
+    Also handles the installing and updating of the server in cache.
+
+    initialize_resources() needs to be called during (or after) plugin_loaded() for paths to be valid.
+    """
+
+    def __init__(self, package_name, server_directory, server_binary_path):
+        self._initialized = False
+        self._package_name = package_name
+        self._server_directory = server_directory
+        self._server_binary_path = server_binary_path
+        self._package_cache_path = None
```

Comment:
I wondered why `self._package_cache_path` is None
I tried setting this line to be:
`self._package_cache_path = os.path.join(sublime.cache_path(), self._package_name)`
But if I do that I get an Error in the console. 
PermissionError: [Errno 13] Permission denied: 'LSP-eslint'

So I guess this can stay this way. :)

Reply:
`sublime.cache_path()` will only return valid (absolute) path when called from `plugin_loaded` (or after, of course) so I can't use it immediately from the constructor.